File: backend/app/core/impersonation.py
"""Brand impersonation detection (F12 signal 1 & look-alike brands).

Detects when an app uses a name or package mimicking a known brand but does not
originate from the brand's official package list.
"""

from __future__ import annotations

from app.core.rules_loader import Brand, Rules, get_rules
from app.models.schemas import ImpersonationResult


def detect_impersonation(
    label: str | None,
    package: str | None,
    rules: Rules | None = None,
) -> ImpersonationResult | None:
    """Checks if an app impersonates a known brand.

    If an alias of a brand appears in the label or package name, but the package
    is NOT in the brand's official packages list, impersonation is detected.
    """
    rules = rules or get_rules()
    target_text = f"{label or ''} {package or ''}".lower()
    pkg_clean = (package or "").lower()

    for brand in rules.brands:
        # Check if any brand alias appears in label or package
        matched_alias = False
        for alias in brand.aliases:
            if alias in target_text:
                matched_alias = True
                break

        if matched_alias:
            # If the brand has known official packages
            if brand.official_packages:
                if pkg_clean not in brand.official_packages:
                    return ImpersonationResult(
                        brand=brand.name,
                        reason_key="impersonation.detected",
                    )

    return None

File: shared/src/commonMain/kotlin/com/appautopsy/analysis/core/Impersonation.kt
package com.appautopsy.analysis.core

import com.appautopsy.analysis.model.ImpersonationResult
import com.appautopsy.analysis.rules.Rules

/**
 * Brand impersonation detection.
 *
 * An app wearing a known brand's name or package while not being that brand's
 * package is claiming to be something it is not. Only brands with a known
 * official-package list can be judged — with no list, there is nothing honest
 * to say, so no result.
 */
object Impersonation {

    fun detect(
        label: String?,
        packageName: String?,
        rules: Rules,
    ): ImpersonationResult? {
        val targetText = "${label.orEmpty()} ${packageName.orEmpty()}".lowercase()
        val pkgClean = packageName.orEmpty().lowercase()

        for (brand in rules.brands) {
            val matchedAlias = brand.aliases.any { it in targetText }
            if (!matchedAlias) continue

            if (brand.officialPackages.isNotEmpty() && pkgClean !in brand.officialPackages) {
                return ImpersonationResult(
                    brand = brand.name,
                    reasonKey = "impersonation.detected",
                )
            }
        }

        return null
    }
}
